framework/include/constraints/NodeElemConstraintBase.h

```diff
@@ -0,0 +1,125 @@
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+// MOOSE includes
+#include "Constraint.h"
+#include "NeighborCoupleableMooseVariableDependencyIntermediateInterface.h"
+
+namespace libMesh
+{
+template <typename T>
+class SparseMatrix;
+}
+
+/**
+ * A NodeElemConstraintBase is used when you need to create constraints between
+ * a secondary node and a primary element. It works by allowing you to modify the
+ * residual and jacobian entries on the secondary node and the primary element.
+ */
+class NodeElemConstraintBase
+  : public Constraint,
+    public NeighborCoupleableMooseVariableDependencyIntermediateInterface,
+    public NeighborMooseVariableInterface<Real>
+{
+public:
+  static InputParameters validParams();
+
+  NodeElemConstraintBase(const InputParameters & parameters);
+  ~NodeElemConstraintBase();
+
+  /// Compute the value the secondary node should have at the beginning of a timestep.
+  void computeSecondaryValue(NumericVector<Number> & current_solution);
+
+  /**
+   * Whether or not this constraint should be applied.
+   * @return bool true if this constraint is active, false otherwise
+   */
+  virtual bool shouldApply() { return true; }
+
+  /**
+   * Whether or not the secondary's residual should be overwritten.
+   * @return bool When this returns true the secondary's residual as computed by the constraint will
+   * _replace_ the residual previously at that node for that variable.
+   */
+  virtual bool overwriteSecondaryResidual() const;
+
+  /**
+   * Whether or not the secondary's Jacobian row should be overwritten.
+   * @return bool When this returns true the secondary's Jacobian row as computed by the constraint
+   * will _replace_ the residual previously at that node for that variable.
+   */
+  virtual bool overwriteSecondaryJacobian() const { return overwriteSecondaryResidual(); };
+
+  /**
+   * The variable on the primary elem.
+   * @return MooseVariable & a reference to the primary variable
+   */
+  virtual MooseVariable & primaryVariable() const { return _primary_var; }
+
+  /**
+   * The variable number that this object operates on.
+   */
+  const MooseVariable & variable() const override { return _var; }
+
+protected:
+  /// Compute the value the secondary node should have at the beginning of a timestep.
+  virtual Real computeQpSecondaryValue() = 0;
+
+  /// secondary node variable
+  MooseVariable & _var;
+  /// Primary side variable
+  MooseVariable & _primary_var;
+
+  /// secondary block id
+  unsigned short _secondary;
+  /// primary block id
+  unsigned short _primary;
+
+  /// current node being processed
+  const Node * const & _current_node;
+  /// current element being processed
+  const Elem * const & _current_elem;
+
+  /// Shape function on the secondary side.
+  VariablePhiValue _phi_secondary;
+  /// Shape function on the secondary side.  This will always only have one entry and that entry will always be "1"
```

Comment:
I think this is only true for lagrange variables?

Reply:
It is the shape function for a single node that is the secondary "surface".  It is not element connected to the node.

Reply:
yes but that shape function is set to be 1 because that's the value of the lagrange shape function at that node

Reply:
This is hardcoding the shape function for a node.  I think this has more to do with setting the secondary_primary lagrange multiplier than the type of shape function.  I don't think this is right for elements with dofs that aren't located at physical coordinates, I don't know what that would do.  This was already in the code before I started this PR to add AD so this is the behavior people are currently using.

Reply:
yeah I m thinking we should be limiting the FE families allowed for the variables here. ANother PR